Issue #1065 Add cc/bcc logic

michael-volynets opened the pull request:
Issue #1065 

Here I added the JS logic to send CC/BCC messages,

Things to do:

- [x] Collapse inputs
- [x] Add CC/BCC buttons in the top-right corner
- [x] Drag & Drop
- [x] Save CC/BCC recipients in Draft
- [x] Fix&Write tests
- [x] Fix CSS in reply window

tomholub commented on extension/js/common/composer/interfaces/composer-types.ts:
```diff
@@ -2,12 +2,18 @@
-  element: HTMLElement
+export type SendingType = 'to' | 'cc' | 'bcc';
+
+export interface BaseRecipient {
```
Why not `Recipients = {to: string[], cc?: string[], bcc?: string[]}` ? That will simplify your code

michael-volynets replied:
Sorry, but I don't exactly understand how it will simplify the code.

Could you please explain to me a little bit more what do you mean?

tomholub replied:
You will be able to remove this whole chunk of code:
```js
    const to: string[] = [];
    const cc: string[] = [];
    const bcc: string[] = [];
    for (const recipient of recipients) { // Using JS loop to not dublicate .filter().map()
      switch (recipient.sendingType) {
        case 'to':
          to.push(recipient.email);
          break;
        case 'cc':
          cc.push(recipient.email);
          break;
        case 'bcc':
          bcc.push(recipient.email);
          break;
      }
    }
```
createMsgObj API will become more obvious, from:
```js
await Google.createMsgObj(acctEmail, acctEmail, [{ email: acctEmail, sendingType: 'to' }], subj, { 'text/html': emailMsg }, emailAtts);
```
to:
```js
await Google.createMsgObj(acctEmail, acctEmail, { to: [acctEmail] }, subj, { 'text/html': emailMsg }, emailAtts);
```

michael-volynets replied:
Okay, got it,
If it's the only reason so I can just change the code this way:
```TS
    const result: Recipients = { to: [], cc: [], bcc: [] };
    // tslint:disable-next-line: standard-loops
    recipients.forEach(r => result[r.sendingType].push(r.email));
```

Because there are some places where it's better to use `BaseRecipient` or `RecipientElement` like here:

if we change method signature from this:
```TS
  public static createMsgObj = async (
    acctEmail: string, from: string = '', recipientElements: RecipientElement[] = [], subject: string = '', by: SendableMsgBody, atts?: Att[], threadRef?: string
  ): Promise<SendableMsg> 
```
to that:
```TS
  public static createMsgObj = async (
    acctEmail: string, from: string = '', recipients: Recipients, subject: string = '', by: SendableMsgBody, atts?: Att[], threadRef?: string
  ): Promise<SendableMsg> 
```
Then we need to join (`Array.join`) all `to:`, `cc:` and `bcc:` values to make this validation working. (I guess I will need to do these joins in other places too).
```TS
    if (!recipientElements.length) {
      throw new Error('The To: field is empty. Please add recipients and try again');
    }
    const invalidEmails = recipientElements.filter(r => !Str.isEmailValid(r.email));
    if (invalidEmails.length) {
      throw new Error(`The To: field contains invalid emails: ${invalidEmails.join(', ')}\n\nPlease check recipients and try again.`);
    }
```

So, I would like to leave it as is but change that large `for` cycle to the small one I have written above.

tomholub replied:
I'll take a look later today

tomholub replied:
In `if (!recipientElements.length) {` you should actually do `recipients.to.length`. Email without `To` header is invalid (to my knowledge). So you would, as it is now, have to rewrite that with a filter anyway.

Maybe you meant `Array.concat` ? You can do `Array.prototype.concat(...Object.values(recipients))`. If there are more than two places that need this, you could put it as ComposerRecipients (or whatever is the class name) method, etiher normal or static, eg `asArray(recipients: Recipients): string[]` or something like that.

I didn't examine the code in deep detail so I may be missing something. I just feel like `(acctEmail, acctEmail, [{ email: acctEmail, sendingType: 'to' }]` is a bit awkward as an API.

michael-volynets replied:
Okay, got it, 
Yes, it makes sense, I will rewrite it according to what you just wrote.

michael-volynets replied:
It should be easier to use  :+1:

tomholub replied:
That's always nice :)

michael-volynets replied:
It isn't resolved yet

I will resolve it when I finish all other things

michael-volynets commented:
@tomholub I want to test it on Firefox Browser but I get this error when I run the command `run_firefox`:
![image](https://user-images.githubusercontent.com/49104848/63278816-a896d280-c2b0-11e9-9a63-bcafddcfa470.png)

I checked if there is the Firefox profile and everything is okay,
I didn't have that issue before.

Does web-ext work well on your side?

tomholub commented:
Yes, I use it all the time. I would try to rename the profile folder, then try to re-run it. It may create another profile in its place, and run just fine (maybe?).

tomholub commented:
On linux

michael-volynets commented:
Okay, I will check it,
It worked before, I don't know what's happened

Thanks.

michael-volynets commented:
@tomholub I think you can start reviewing it,
The functionality is almost done, I listed things that aren't done in the PR header.

tomholub commented:
Thanks! I'll take a look

michael-volynets commented:
@tomholub I have a question,

When I'm replying to a message can I somehow extract CC & BCC headers? 
I found that we parse `To` Header from the HTML page and I can extract CC/BCC the same way but they are in the popover:
![image](https://user-images.githubusercontent.com/49104848/63432456-dc493800-c429-11e9-9147-9db7df9d0bac.png)
and I find little difficult to parse it and maybe parsing is not the best way for doing that.

Maybe you know some other ways to extract that data?

tomholub commented on extension/js/common/composer/composer-contacts.ts:
```diff
@@ -623,7 +623,7 @@ export class ComposerContacts extends ComposerComponent {
     }
     const restHTML = `<span class="rest"><span id="rest_number"></span> others</span>`;
-    container.html(restHTML);
+    container.html(restHTML); // xss-direct
```
Yup, correct

michael-volynets commented:
Also, I've got an issue with tests,

I added `data-test='collapsed'` attribute to the element that hides inputs.
```HTML
          <div class="collapsed" data-test="collapsed">
            <span class="placeholder">Recipients</span>
            <div class="email_preview"></div>
          </div>
``` 
And when I want to click on it, it seems to not be invoked (I added some logs in the methods to check and I couldn't see them when the test was running).
This code `await composePageOrFrame.click('@collapsed');` should open inputs so I can type the recipients' emails but it doesn't work. Do you know why it may not work?

tomholub commented:
You should call it in line with other ones, like `action-expand-cc-bcc-fields`

But that's not causing the issue. I'm not sure what is.

michael-volynets commented:
Okay, I will rename it and try to investigate why that happens

tomholub commented:
You could try to `waitAny` it, then get the element and do `await el.click()`, as a way to debug it.

You could also execute javascript on the frame, with `$(".yourclass").click()` in it as a test, through puppeteer.

tomholub commented:
Other than the send_from issue, please also align cc/bcc to the right side to have the same distance to the right as the input field has from the left:

![image](https://user-images.githubusercontent.com/6306961/63455322-16bbd080-c43c-11e9-8027-efdc9f048221.png)

Currently it's not clear how was the distance from the right side chosen :)

![image](https://user-images.githubusercontent.com/6306961/63455356-29cea080-c43c-11e9-8b8f-81f30b3b6237.png)

Please also make it lowercase: `cc`, `bcc`.

tomholub commented:
> @tomholub I have a question,
> 
> When I'm replying to a message can I somehow extract CC & BCC headers?
> I found that we parse `To` Header from the HTML page and I can extract CC/BCC the same way but they are in the popover:
> ![image](https://user-images.githubusercontent.com/49104848/63432456-dc493800-c429-11e9-9147-9db7df9d0bac.png)
> and I find little difficult to parse it and maybe parsing is not the best way to doing that.
> 
> Maybe you know some other ways to extract that data?

Sorry, I missed that. Don't parse it from dom. Get it from Gmail API. We are already getting this info from Gmail, I'll find where.

tomholub commented:
![image](https://user-images.githubusercontent.com/6306961/63533939-631b1500-c4fd-11e9-83ad-4ac4d561bcb7.png)

Here. Just change:
```diff
-if (!isReplyBox || (threadId && threadId !== threadMsgId && to.length && subject)) {
+if (!isReplyBox || !threadId) {
   return; // either not a reply box, or reply box & has all needed params
 }
```
That means we will be ignoring the `to` and `subject` in url, and always fetching metadata from Gmail if it's a reply.

Later we will get rid of these header url parameters altogether (except for threadId), but in another PR (this PR is already getting big).

In the `gmailMsg` object you will find the needed headers.

michael-volynets commented:
Thanks for the previous answer.

Also, Is there any reason why you added this CSS class:
```CSS
table#compose td.input.show_send_from {
    height: 84px;
}
```
That code adds these paddings:
![image](https://user-images.githubusercontent.com/49104848/63571184-29dbb700-c588-11e9-9902-547b784a248e.png)

But without that paddings it looks like this:
![image](https://user-images.githubusercontent.com/49104848/63571208-38c26980-c588-11e9-8d9e-ee2abdcc4c7e.png)

In my opinion, it looks better without those paddings and I want to remove that CSS code, what do you think?

tomholub commented:
The goal of the css, I think, was to make all of these 3 cells the same height:
 - recipients
 - sender
 - subject

I tried to measure it and it's already a bit broken (prod):

![image](https://user-images.githubusercontent.com/6306961/63579163-82f41c80-c581-11e9-982a-a3332a07b25a.png)

Which explains the broken padding, visually (supposed to be the same distance):

![image](https://user-images.githubusercontent.com/6306961/63579203-9bfccd80-c581-11e9-96a7-3188c9daf881.png)

If that was made the same distance, it will all work:

![image](https://user-images.githubusercontent.com/6306961/63579296-d23a4d00-c581-11e9-88ac-efc11d432827.png)

For the one that you proposed, I'm lacking any sense of consistency, don't know what to make of it:

![image](https://user-images.githubusercontent.com/6306961/63579451-3bba5b80-c582-11e9-9ba7-c68298522792.png)

So from my point of view, as it is now in production is slightly broken, your "before" screenshot is somewhat more broken, and your proposal also :) I'm not a UI designer (as you can tell) but I really have a thing for UIs with apparent visual consistency (especially padding / sizing / positioning).

michael-volynets commented:
Got the goal of that code, Thanks!

I will fix the CSS to make the paddings the same in all 3 inputs (Recipients&From&Subject).

tomholub commented:
If you will be editting it, here is I think the best form of consistency, which is a bit more compact:

![image](https://user-images.githubusercontent.com/6306961/63580685-e16eca00-c584-11e9-9516-6f9e1d0cd6a4.png)

Remember to make sure it's consistent regardless is "from" is showing or not.

michael-volynets commented:
Looks great, 

Got it

michael-volynets commented:
@tomholub Could you please review it again?

I have done all your requested changes,

Thanks

tomholub commented:
Some user:

![image](https://user-images.githubusercontent.com/6306961/63652343-da93b300-c74e-11e9-9af4-ee4428a9cbc4.png)

tomholub commented:
Sorry, I somehow missed the message above. I'll take another look

tomholub commented:
I have to block merging for now for lack of visual consistency :laughing: ... sorry I'm quite anal about these things :roll_eyes: 

![image](https://user-images.githubusercontent.com/6306961/63652627-da48e700-c751-11e9-91cf-6611bb5b63ed.png) means good, ![image](https://user-images.githubusercontent.com/6306961/63652634-e5037c00-c751-11e9-8771-2e3ba5764e54.png) means off by 1 pixel (should still be fixed), ![image](https://user-images.githubusercontent.com/6306961/63652640-f3519800-c751-11e9-9257-9f62942ea3c6.png) means too far off

1) Outer border padding. All of the red lines are supposed to be the same distance, but they all seem to differ.

![image](https://user-images.githubusercontent.com/6306961/63652712-db2e4880-c752-11e9-8fa0-bfe46e1f2299.png)

2) vertically it seems to need a bit more space. When I measured it, it's just 1px, but my eye is maybe really missing that pixel :)

![image](https://user-images.githubusercontent.com/6306961/63652692-a6ba8c80-c752-11e9-80e7-e07548a3c1ec.png)

3) cc and bcc have inconsistent margin from the other elements, vertically

![image](https://user-images.githubusercontent.com/6306961/63652773-a79fee00-c753-11e9-8f62-5758ad0cdac5.png)

tomholub commented on extension/chrome/elements/reply_pubkey_mismatch.ts:
```diff
@@ -103,10 +104,15 @@ Catch.try(async () => {
     const body = { 'text/plain': $('#input_text').get(0).innerText };
-    const message = await Google.createMsgObj(acctEmail, from, to, subject, body, [att], threadId);
+    const recipients: BaseRecipient[] = to.map(email => {
+      const recipient: BaseRecipient = { email, sendingType: 'to' };
```
Please use Recipients instead of BaseRecipient[]

tomholub commented on extension/chrome/settings/modules/backup.ts:
```diff
@@ -218,7 +218,7 @@ Catch.try(async () => {
     const emailMsg = String(await $.get({ url: '/chrome/emails/email_intro.template.htm', dataType: 'html' }));
     const emailAtts = [asBackupFile(acctEmail, armoredKey)];
-    const msg = await Google.createMsgObj(acctEmail, acctEmail, [acctEmail], GMAIL_RECOVERY_EMAIL_SUBJECTS[0], { 'text/html': emailMsg }, emailAtts);
+    const msg = await Google.createMsgObj(acctEmail, acctEmail, [{ email: acctEmail, sendingType: 'to' }], GMAIL_RECOVERY_EMAIL_SUBJECTS[0], { 'text/html': emailMsg }, emailAtts);
```
Please use Recipients here

tomholub commented on extension/js/common/api/email_provider_api.ts:
```diff
@@ -9,28 +9,32 @@ import { Value, Dict, Str } from '../core/common.js';
 
   public static createMsgObj = async (
-    acctEmail: string, from: string = '', to: string[] = [], subject: string = '', by: SendableMsgBody, atts?: Att[], threadRef?: string
+    acctEmail: string, from: string = '', recipientElements: BaseRecipient[] = [], subject: string = '', by: SendableMsgBody, atts?: Att[], threadRef?: string
```
Please use Recipients here

tomholub commented on extension/js/common/composer.ts:
```diff
@@ -970,7 +978,7 @@ export class Composer {
         // Firefox will not always respond to initial automatic $input_text.blur()
         // Recipients may be left unrendered, as standard text, with a trailing comma
-        await this.composerContacts.parseRenderRecipients(this.S.cached('input_to_container')); // this will force firefox to render them on load
+        this.composerContacts.parseRenderRecipients(this.S.cached('input_to')).catch(Catch.reportErr); // this will force firefox to render them on load
```
Any reason not to await this? Are there external API calls?

michael-volynets replied:
Sorry, I think I added it accidentally because it could block the thread.

I tested it along with `await` and it works the same way, thanks!

tomholub commented:
By the way I like this :+1:

![image](https://user-images.githubusercontent.com/6306961/63652912-8b9d4c00-c755-11e9-8cb2-8a87ac33cf1b.png)

It wasn't part of the issue, but I think it increases usability. I'll go ahead and adjust the design a bit.

tomholub commented:
Another issue: it seems to create broken html?
```html
<span class="email_address has_pgp" "="">
```

tomholub commented:
Sometimes I click away from the recipients and it does not result in showing recipient preview. I'll make a video.

tomholub commented:
When I clicked away from input, the clicked back, it stopped responding to input / stopped adding email bubbles! This one is quite serious, if you can reproduce it, please also add a test for this before fixing it.

![image](https://user-images.githubusercontent.com/6306961/63653215-71656d00-c759-11e9-9982-0b2d8730efc8.png)

I keep clicking back and forth, pressing tab, etc, the last two emails do not get processed.

tomholub commented:
It's definitely on the right track, but has many quirks that need fixing - see all above.

michael-volynets commented:
https://github.com/FlowCrypt/flowcrypt-browser/pull/1939#issuecomment-524644262

Thanks, I that was my typo mistake

michael-volynets commented:
https://github.com/FlowCrypt/flowcrypt-browser/pull/1939#issuecomment-524644479
Fixed,

Now when you will focus RecipientElement and then blur it the inputs will collapse

michael-volynets commented:
> When I clicked away from input, the clicked back, it stopped responding to input / stopped adding email bubbles! This one is quite serious, if you can reproduce it, please also add a test for this before fixing it.
> 
> ![image](https://user-images.githubusercontent.com/6306961/63653215-71656d00-c759-11e9-9982-0b2d8730efc8.png)
> 
> I keep clicking back and forth, pressing tab, etc, the last two emails do not get processed.
> 
> Tried hitting enter, and it processed them both as one email, even though there is a space between them:
> 
> ![image](https://user-images.githubusercontent.com/6306961/63653235-bf7a7080-c759-11e9-91fa-93a439a59953.png)

Sorry, but we didn't have that functionality before, we can only add 2 or more emails at one time when we write them by comma
e.g. `test@test.com,test2@test.com` then they will be processed.

But if you noticed it I will add it, it won't take a lot of time to write that.

michael-volynets commented:
Little question, I found this code: 
```TS
  private getReplyParams = (convoRootEl: JQueryEl): FactoryReplyParams => {
    const headers = Google.determineReplyCorrespondents(this.acctEmail, this.addresses, {
      lmSender: this.domGetMsgSender(convoRootEl),
      lmRecipients: this.domGetMsgRecipients(convoRootEl),
      lmReplyTo: undefined, // cannot get this info from DOM, will be later added from API
    });
....
  }
```
In `gmail_element_replacer.ts` file.
Can I get rid of this? I mean I want to replace it with Gmail API call to fetch the to/cc/bcc data, will it be okay?

michael-volynets commented:
> Little question, I found this code:
> 
> ```ts
>   private getReplyParams = (convoRootEl: JQueryEl): FactoryReplyParams => {
>     const headers = Google.determineReplyCorrespondents(this.acctEmail, this.addresses, {
>       lmSender: this.domGetMsgSender(convoRootEl),
>       lmRecipients: this.domGetMsgRecipients(convoRootEl),
>       lmReplyTo: undefined, // cannot get this info from DOM, will be later added from API
>     });
> ....
>   }
> ```
> 
> In `gmail_element_replacer.ts` file.
> Can I get rid of this? I mean I will only use Gmail API call (in composer.ts) to fetch the to/cc/bcc data, will it be okay?

Resolved

michael-volynets commented:
@tomholub, It's ready for review,

One thing I have done without asking you is I removed this arrow in the top-right corner:
![image](https://user-images.githubusercontent.com/49104848/64080989-671c1380-cd03-11e9-8bed-f127116c7e01.png)

because it brought me some problems with CSS and I didn't want to spend a lot of time on that.
Now it immediately opens emails aliases in the reply box (the same way like you open a compose window from the button in Gmail sidebar). If it isn't okay for you (if you want to leave that arrow) so I can somehow try to fix that.

tomholub commented:
> Now it immediately opens emails aliases in the reply box (the same way like you open a compose window from the button in Gmail sidebar). If it isn't okay for you (if you want to leave that arrow) so I can somehow try to fix that.

It was done this way because reply box has much more limited vertical space. I'll try it.

tomholub commented on extension/js/common/composer/composer-draft.ts:
```diff
@@ -67,12 +67,6 @@ export class ComposerDraft extends ComposerComponent {
         await this.abortAndRenderReplyMsgComposeTableIfIsReplyBox('!armored');
         return false;
       }
-      if (String(parsedMsg.headers.subject)) {
```
Removing this introduced a bug, subjects are not loaded when I load a draft.

tomholub commented on extension/js/common/composer/composer-draft.ts:
```diff
@@ -113,7 +107,13 @@ export class ComposerDraft extends ComposerComponent {
-        const mimeMsg = await Mime.encode(body, { To: to.join(','), From: this.composer.getSender(), Subject: subject }, []);
+        const recipients: Recipients = { to: [], cc: [], bcc: [] };
+        // tslint:disable-next-line: standard-loops
+        this.composer.getRecipients().forEach(r => recipients[r.sendingType]!.push(r.email));
```
standard for/of loop can be used here

tomholub replied:
That's why we have a linter for it! :)

tomholub commented on extension/js/common/composer/composer-draft.ts:
```diff
@@ -173,38 +173,30 @@ export class ComposerDraft extends ComposerComponent {
     if (typeof passphrase !== 'undefined') {
       const result = await PgpMsg.decrypt({ kisWithPp: await Store.keysGetAllWithPp(this.urlParams.acctEmail), encryptedData: Buf.fromUtfStr(encryptedArmoredDraft) });
       if (result.success) {
+        if (headers.subject) {
```
This looks like it should work, but for whatever reason, when tested, subject is not added. Not sure why

tomholub commented on extension/js/common/core/mime.ts:
```diff
@@ -111,20 +114,16 @@ export class Mime {
-        if (to.address) {
-          headerTo.push(String(to.address));
+  private static headerGetAddress = (parsedMimeMsg: MimeContent, headersNames: Array<SendingType | 'from'>) => {
+    const result: { from: string[], to: string[], cc: string[], bcc: string[] } = { from: [], to: [], cc: [], bcc: [] };
```
Why would from be an array?

michael-volynets replied:
I agree, it would be better to have `from` like `string | undefined`.

I wanted to have all headers the same type so I can work with them easier

tomholub commented on extension/js/common/core/mime.ts:
```diff
@@ -111,20 +114,16 @@ export class Mime {
+  private static headerGetAddress = (parsedMimeMsg: MimeContent, headersNames: Array<SendingType | 'from'>) => {
+    const result: { from: string[], to: string[], cc: string[], bcc: string[] } = { from: [], to: [], cc: [], bcc: [] };
+    for (const hdrName of headersNames) {
+      if (Array.isArray(parsedMimeMsg.headers[hdrName])) {
```
Hopefully `from` should never be an array? So not sure if this would work?

tomholub commented on test/source/mock.ts:
```diff
@@ -12,7 +12,7 @@ export const mock = async (logger: (line: string) => void) => {
     }
     const filename = `${email.replace(/[^a-z0-9]+/g, '')}.json`;
-    const url = `https://github.com/FlowCrypt/flowcrypt-bin/blob/master/gmail-mock-data/${filename}?raw=true`;
+    const url = `https://github.com/michael-volynets/flowcrypt-bin/blob/update-mock-data/gmail-mock-data/${filename}?raw=true`; // UNDO: before commit
```
Note for me - remember to fix
